packages/service-library/tests/test_async_utils.py

```diff
@@ -0,0 +1,150 @@
+import asyncio
+import copy
+import random
+from collections import deque
+from time import time
+from typing import Any, Dict, List
+
+import pytest
+
+from servicelib.async_utils import run_serialized_in_context
+
+
+class LockedStore:
+    __slots__ = ("_queue", "_lock")
+
+    def __init__(self):
+        self._queue = deque()
+        self._lock = asyncio.Lock()
+
+    async def push(self, item: Any):
+        async with self._lock:
+            self._queue.append(item)
+
+    async def get_all(self) -> List[Any]:
+        async with self._lock:
+            return list(self._queue)
+
+
+async def test_context_aware_dispatch() -> None:
+    @run_serialized_in_context(target_args=["c1", "c2", "c3"])
+    async def orderly(c1: Any, c2: Any, c3: Any, control: Any) -> None:
+        _ = (c1, c2, c3)
+        sleep_interval = random.uniform(0, 0.01)
+        await asyncio.sleep(sleep_interval)
+
+        context = dict(c1=c1, c2=c2, c3=c3)
+        await locked_stores[make_key_from_context(context)].push(control)
+
+    def make_key_from_context(context: Dict) -> str:
+        return ".".join([f"{k}:{v}" for k, v in context.items()])
+
+    def make_context():
+        return dict(
+            c1=random.randint(0, 10), c2=random.randint(0, 10), c3=random.randint(0, 10)
+        )
+
+    contexts = [make_context() for _ in range(10)]
+
+    locked_stores = {}
+    expected_outcomes = {}
+    for context in contexts:
+        key = make_key_from_context(context)
+        locked_stores[key] = LockedStore()
+        expected_outcomes[key] = deque()
+
+    tasks = deque()
```

Comment:
I don't know anybody that loves so much ``deque`` :-)

Reply:
I prefer avoiding slow queues ;-)